### PR TITLE
Implement the Re-rebalance coherence RFC

### PR DESCRIPTION
**Pull request by weiznich**
This is the first time I touch anything in the compiler so just tell me if I got something wrong.

Big thanks to @sgrif for the pointers where to look for those things.
cc #55437

### CODE REVIEW

**Review by varkor — CHANGES_REQUESTED** [review]
I haven't checked over the tests yet, but the code looks good. I'll try to look at the tests soon.

File: src/doc/unstable-book/src/language-features/re-rebalance-coherence.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+[#55437]: https://github.com/rust-lang/rust/issues/55437
+
+------------------------
+
+The `re_rebalance_coherence` feature tweaks the rules which trait
+impls are allowed in crates.
+The following rule is used:
+
+Given `impl<P1..=Pn> Trait<T1..=Tn> for T0`, an impl is valid only if at
```
[the remark this answers is not in the document]
Comment (weiznich) [reply src/doc/unstable-book/src/language-features/re-rebalance-coherence.md]: That's directly copied from the RFC. I think it should be "if and only if" but's not written there. (cc @sgrif who has written that thing)

**Comment by weiznich** [comment]
@varkor I've addressed your review comments.

**Comment by weiznich** [comment]
@varkor I've rebased the PR.

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -393,6 +393,9 @@ declare_features! (
     // `extern` in paths
     (active, extern_in_paths, "1.23.0", Some(55600), None),
 
+    // Use `?` as the Kleene "at most one" operator
```
Comment (varkor) [inline src/libsyntax/feature_gate.rs]: You seem to have accidentally pulled in some extra changes with the rebase.

**Comment by varkor** [comment]
At the moment, the fact that all the coherence tests have been duplicated under the new feature flag seems a little excessive. I'm not quite sure what the correct approach here is though: we do want to make sure we don't break anything. I wonder if it would be appropriate to run the tests in two modes, a bit like NLL does at the moment. Alternatively, we could have a few tests that cover a representative sample of existing cases, under the assumption that the algorithm is mostly the same anyway and when stabilised, the existing tests are going to come into effect anyway.

Could you point out which tests are new (and demonstrate the new behaviour), so I can check those in particular?

I think @nikomatsakis will have better ideas about the right solution re. tests, so I'm going to reassign. The actual change looks good to me, though.

r? @nikomatsakis

**Comment by weiznich** [comment]
> Could you point out which tests are new (and demonstrate the new behaviour), so I can check those in particular?

[This one](https://github.com/rust-lang/rust/pull/56145/files#diff-5aaa46e5eb05735e9d603b98a850ee14) and [this one](https://github.com/rust-lang/rust/pull/56145/files#diff-6b1f9dd40f86d9d6559fddc781f2ec13) is new. All other tests are just a copied version of the old coherence tests with the `re_rebalance` feature enabled.

> At the moment, the fact that all the coherence tests have been duplicated under the new feature flag seems a little excessive. I'm not quite sure what the correct approach here is though: we do want to make sure we don't break anything. I wonder if it would be appropriate to run the tests in two modes, a bit like NLL does at the moment. Alternatively, we could have a few tests that cover a representative sample of existing cases, under the assumption that the algorithm is mostly the same anyway and when stabilised, the existing tests are going to come into effect anyway.

I'm not sure what would be a representative sample of the existing test cases here, nor I'm knowing what needs to be changed to run the existing tests using two modes. Would be great to get some more input here.

**Comment by weiznich** [comment]
I'm not sure why those 2 tests fail. The corresponding `.stderr` files containing exactly those error messages...

**Comment by weiznich** [comment]
@varkor @nikomatsakis Any news here?

**Review by nikomatsakis — APPROVED** [review]
Seems great! I would like before r+'ing though to spend a bit more time checking over the tests.

**Comment by nikomatsakis** [comment]
I forgot about this question that @varkor raised:

> At the moment, the fact that all the coherence tests have been duplicated under the new feature flag seems a little excessive.

One option here is to rewrite the duplicated tests using `revisions`. The idea is to include a comment like:

```
// revisions: old new
```

This will cause the same test to be built twice, once with `#[cfg(old)]` and once with `#[cfg(new)]` in scope.

You can then do things like:

```rust
#![cfg_attr(new, feature(new_algorithm))]
```

@weiznich -- do you want to take a shot at collapsing the tests this way?

**Comment by weiznich** [comment]
@nikomatsakis I will try that. What's the right strategy [here](https://github.com/rust-lang/rust/pull/56145/files#diff-cee15537dc425f257a8d9061279cd500R13)? In that case there are already revisions.

File: src/test/ui/re_rebalance_coherence/coherence-cow.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(re_rebalance_coherence)]
+
+// revisions: a b c
```
Comment (nikomatsakis) [inline src/test/ui/re_rebalance_coherence/coherence-cow.rs]: Well, in this case you would have to add another set of three revisions, I suppose. e.g., a, re_a, b, re_b, c, re_c

**Pull request merged**